s3fs/core.py

```diff
@@ -941,126 +1064,125 @@ def copy_managed(self, path1, path2, block=5 * 2**30, **kwargs):
-                        UploadId=mpu['UploadId'],
-                        CopySource=path1,
-                        CopySourceRange=brange)
-                    break
-                except S3_RETRYABLE_ERRORS as exc:
-                    if attempt < self.retries:
-                        logger.debug('Exception %r on S3 write, retrying', exc,
-                                     exc_info=True)
-                    time.sleep(1.7 ** attempt * 0.1)
-                except Exception as exc:
-                    raise IOError('Write failed: %r' % exc) from exc
-            parts.append({'PartNumber': i + 1,
-                          'ETag': out['CopyPartResult']['ETag']})
-        self._call_s3(
+            Bucket=bucket, Key=key, **kwargs
+        )
+        # attempting to do the following calls concurrently with gather causes
+        # occasional "upload is smaller than the minimum allowed"
+        out = [
+            await self._call_s3(
+                self.s3.upload_part_copy,
+                Bucket=bucket,
+                Key=key,
+                PartNumber=i + 1,
+                UploadId=mpu['UploadId'],
+                CopySource=path1,
+                CopySourceRange="bytes=%i-%i" % (brange_first, brange_last))
+            for i, (brange_first, brange_last) in enumerate(_get_brange(size, block))
+        ]
+        parts = [{'PartNumber': i + 1,
+                 'ETag': o['CopyPartResult']['ETag']} for i, o in enumerate(out)]
+        await self._call_s3(
             self.s3.complete_multipart_upload,
             Bucket=bucket,
             Key=key,
             UploadId=mpu['UploadId'],
             MultipartUpload={'Parts': parts})
         self.invalidate_cache(path2)
 
-    def copy(self, path1, path2, **kwargs):
+    async def _cp_file(self, path1, path2, **kwargs):
         gb5 = 5 * 2**30
         path1 = self._strip_protocol(path1)
-        size = self.info(path1)['size']
+        bucket, key, vers = self.split_path(path1)
+        size = (await self._info(path1, bucket, key, version_id=vers))['size']
         if size <= gb5:
             # simple copy allowed for <5GB
-            self.copy_basic(path1, path2, **kwargs)
+            await self._copy_basic(path1, path2, **kwargs)
         else:
             # serial multipart copy
-            self.copy_managed(path1, path2, **kwargs)
-
-    def bulk_delete(self, pathlist, **kwargs):
+            await self._copy_managed(path1, path2, size, **kwargs)
+
+    async def _clear_multipart_uploads(self, bucket):
+        """Remove any partial uploads in the bucket"""
+        out = await self._call_s3(self.s3.list_multipart_uploads, Bucket=bucket)
+        await asyncio.gather(*[
+            self._call_s3(self.s3.abort_multipart_upload,
+                          Bucket=bucket,
+                          Key=upload['Key'],
+                          UploadId=upload['UploadId'])
+            for upload in out['Contents']
+        ])
```

Comment:
I've run into situations before where AWS will cry foul if I do too many operations at a time.  If this patterns repeats itself frequently then we might consider some sort of `batched_gather` that runs at most 20 at a time or something.

Reply:
I am thinking that the exponential backoff retries are enough. I have no trouble pipe-ing and cat-ing 1300 small files in one go.

Reply:
Ah!  So each of the sub-functions have their own backoff strategy.  Neat.

Reply:
Anything that goes through _call_s3, yes. There may still be a couple of exceptions to this, but fewer than before, so this rewrite has helped with that.